### DISCUSSION TITLE
How can I add "QueryStringParams" that are not url encoded?

### DISCUSSION THREAD

**Discussion by muffl0n** [Orange-OpenSource/hurl#3243] [Q&A] [answered]
My servers are throwing an 400 error when they get requests with a parameter like `foo={bar}`. To test this I currently use curl like this:
```
$ curl 'https://www.example.com/index.html?foo=\{bar\}'
...
> GET index.html?foo={bar} HTTP/2
> Host: www.example.com
> User-Agent: curl/8.10.0
> Accept: */*
>
* Request completely sent off
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
< HTTP/2 400
...
HTTP/2 400
...
```
Unfortunately, when I try to test this with hurl
```
GET https://www.example.com/index.html

[QueryStringParams]
foo: {bar}

HTTP 200
```
the parameter gets encoded and the servers respond with 200:
```
...
* Request:
* GET https://www.example.com/index.html
* [QueryStringParams]
* foo: {bar}
*
* Request can be run with the following curl command:
* curl 'https://www.example.com/index.html?foo=%7Bbar%7D'
*
...
> GET /index.html?foo=%7Bbar%7D HTTP/2
> Host: www.example.com:443
> Proxy-Connection: Keep-Alive
> Host: www.example.com
> Accept: */*
> User-Agent: hurl/5.0.1
>
* Response: (received 544082 bytes in 1924 ms)
*
< HTTP/1.0 200 Connection Established
< HTTP/2 200
...
```

Is there a way to *not*  encode the parameters defined in `QueryStringParams`?

**Comment by jcamiel** [comment, marked as answer]
Hi @muffl0n 

Parameters in `[QueryStringParams]` section are always url encoded. You could write the URl directly without the query section, but there is a parsing error if we want a request to `https://www.example.com/index.html?foo={bar}`:

```hurl
GET https://www.example.com/index.html?foo={bar}
HTTP 400
```

```shell
$ hurl test.hurl
error: Parsing URL
  --> /tmp/url/url2.hurl:1:44
   |
 1 | GET https://www.example.com/index.html?foo={bar}
   |                                            ^ illegal character <{>
   |
```

The only way I see to do this is to "trick" the parser and use a variable in the URL:

```hurl
GET https://www.example.com/index.html?foo={{var}}
[Options]
variable: var={bar}
HTTP 400
```

It's not pretty but it's working😅 (not that, in this case, the curl logs shown with `hurl --verbose` is not good because `{` are not escaped though)

**Reply by muffl0n** [reply]
Hi @jcamiel, thank you very much for this detailed answer! Does exactly what I was looking for! ❤️

**Reply by jcamiel** [reply]
Yes, note that we should be able to write this kind of file:

```hurl
GET https://www.example.com/index.html?foo={bar}
HTTP 400
```

Without any trick, issue created here => https://github.com/Orange-OpenSource/hurl/issues/3244 you can register to it it you want to track its advancement.

**Reply by muffl0n** [reply]
Awesome! Thank you!
